adafruit_gps.py

```diff
@@ -662,8 +699,7 @@ def _parse_gsv(self, talker: bytes, data: List[str]) -> bool:
                     # been seen for 30 seconds
                     timestamp = time.monotonic()
                     old = []
-                    for i in self.sats:
-                        sat = self.sats[i]
+                    for sat in self.sats.items():
                         if (timestamp - sat[4]) > 30:
```

Comment:
This is a bug! `.items()` always returns a `tuple[k, v]` (length 2) so `sat[4]` will always raise an IndexError, no matter what!

The tests would have caught this, if they were run. 2f6015e for some reason decided to not run them in CI anymore, so regressions like this are likely and will make it into the next release.

<details>

```console
(venv) PS C:\Users\jonas\Downloads\Adafruit_CircuitPython_GPS> pytest
================================================= test session starts =================================================
platform win32 -- Python 3.11.1, pytest-7.3.1, pluggy-1.0.0
rootdir: C:\Users\jonas\Downloads\Adafruit_CircuitPython_GPS
collected 48 items

tests\adafruit_gps_test.py ...............................................F                                      [100%]

====================================================== FAILURES =======================================================
________________________________ test_GPS_update_from_GSV_both_parts_sats_are_removed _________________________________

    def test_GPS_update_from_GSV_both_parts_sats_are_removed():
        gps = GPS(uart=UartMock())
        with mock.patch.object(GPS, "readline") as m:
            with freeze_time("2021-10-20 19:00:00"):
                # first part of the request
                m.return_value = b"$GPGSV,2,1,04,01,40,083,46,02,17,308,41*78\r\n"
                assert gps.update()
                assert gps.total_mess_num == 2
                assert gps.mess_num == 1
                assert gps.satellites == 4
                # first time we received satellites, so this must be None
                assert gps.sats is None
            # some time has passed so the first two satellites will be too old, but
            # this one not
            with freeze_time("2021-10-20 19:00:20"):
                # second part of the request
                m.return_value = b"$GPGSV,2,2,04,12,07,344,39,14,22,228,45*7c\r\n"
                assert gps.update()
                assert gps.total_mess_num == 2
                assert gps.mess_num == 2
                assert gps.satellites == 4
                # we should now have 4 satellites from the two part request
                assert set(gps.sats.keys()) == {"GP1", "GP2", "GP12", "GP14"}

            # time passed (more than 30 seconds) and the next request does not
            # contain the previously seen satellites but two new ones
            with freeze_time("2021-10-20 19:00:31"):
                # a third, one part request
                m.return_value = b"$GPGSV,1,1,02,13,07,344,39,15,22,228,45*7a\r\n"
>               assert gps.update()

tests\adafruit_gps_test.py:467:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
adafruit_gps.py:354: in update
    result = self._parse_gsv(talker, args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <adafruit_gps.GPS object at 0x0000014E020ECB90>, talker = 'GP', data = [1, 1, 2, 13, 7, 344, ...]

    def _parse_gsv(self, talker: bytes, data: List[str]) -> bool:
        # GSV - Satellites in view
        # pylint: disable=too-many-branches

        if data is None or len(data) not in (7, 11, 15, 19):
            return False  # Unexpected number of params.
        data = _parse_data(
            {7: _GSV7, 11: _GSV11, 15: _GSV15, 19: _GSV19}[len(data)],
            data,
        )
        if data is None:
            return False  # Params didn't parse

        talker = str(talker, "ascii")

        # Number of messages
        self.total_mess_num = data[0]
        # Message number
        self.mess_num = data[1]
        # Number of satellites in view
        self.satellites = data[2]

        sat_tup = data[3:]

        satlist = []
        timestamp = time.monotonic()
        for i in range(len(sat_tup) // 4):
            j = i * 4
            value = (
                # Satellite number
                "{}{}".format(talker, sat_tup[0 + j]),
                # Elevation in degrees
                sat_tup[1 + j],
                # Azimuth in degrees
                sat_tup[2 + j],
                # signal-to-noise ratio in dB
                sat_tup[3 + j],
                # Timestamp
                timestamp,
            )
            satlist.append(value)

        if self._sats is None:
            self._sats = []
        for value in satlist:
            self._sats.append(value)

        if self.mess_num == self.total_mess_num:
            # Last part of GSV message
            if len(self._sats) == self.satellites:
                # Transfer received satellites to self.sats
                if self.sats is None:
                    self.sats = {}
                else:
                    # Remove all satellites which haven't
                    # been seen for 30 seconds
                    timestamp = time.monotonic()
                    old = []
                    for sat in self.sats.items():
>                       if (timestamp - sat[4]) > 30:
E                       IndexError: tuple index out of range

adafruit_gps.py:703: IndexError
=============================================== short test summary info ===============================================
FAILED tests/adafruit_gps_test.py::test_GPS_update_from_GSV_both_parts_sats_are_removed - IndexError: tuple index out of range
============================================ 1 failed, 47 passed in 0.26s =============================================
(venv) PS C:\Users\jonas\Downloads\Adafruit_CircuitPython_GPS>
```
</details>

Reply:
Weird, `pre-commit` claims to have run the tests according to the printout from the composite action... it says "Passed" and not "Skipped".

Reply:
I do remember running precommit before merging this, but I don't recall what jobs it ran. Let me dig back in...

Reply:
nvm, it was actually removed in: cca0f0f.

`pre-commit` has nothing to do with running unit tests! if `mypy` was setup via `pre-commit`, it would have caught it.

```console
error: Tuple index out of range  [misc]
```

This part or something similar is needed to run the tests.

https://github.com/adafruit/Adafruit_CircuitPython_GPS/blob/3e41e4494d52b353f0808213c90df75aa146371f/.github/workflows/build.yml#L53-L56

Reply:
I'm big dumb, I'm think of `pylint`.  Yeah, currently `pytest` isn't run though it didn't look like it was set to run before, for what it's worth.  Either way, an oversight on the part of the CI, whoopsies!

I'll create an issue for fixing the libraries in general.  Good catch!

Reply:
Ah, I was 5 mins too late, haha!  I've still created the issue within the repository containing the composite actions.  I think we can add it there so it works for all libraries as opposed to monkeypatching it per library.

Reply:
https://github.com/adafruit/workflows-circuitpython-libs/issues/17

Reply:
Leaving CI aside, I think `.values()` instead of `.items()` fixes this. I can submit a new PR for that

Reply:
no, please don't leave CI aside, it has to be reverted like it was, please run the tests to check before.

I agree, that this code is not very readable. Below is the only way it would work, I guess. But I don't see much added value or readability tbh.

```diff
diff --git a/adafruit_gps.py b/adafruit_gps.py
index 36a17d2..9a3250d 100644
--- a/adafruit_gps.py
+++ b/adafruit_gps.py
@@ -630,10 +630,9 @@ class GPS:
                     # been seen for 30 seconds
                     timestamp = time.monotonic()
                     old = []
-                    for i in self.sats:
-                        sat = self.sats[i]
-                        if (timestamp - sat[4]) > 30:
-                            old.append(i)
+                    for sat, val in self.sats.items():
+                        if (timestamp - val[4]) > 30:
+                            old.append(sat)
                     for i in old:
                         self.sats.pop(i)
                 for sat in self._sats:
```

Reply:
@jkittner maybe you can help with how to run the tests, since adding them to all libs is a larger job than what I can merge with this PR?

Reply:
Ah, it does look like the dict key is being used, so yeah, using `.items()` with unpacking would fix it.  I'm unfamiliar with the inner workings of this library, so unsure whether there is a work around otherwise.

Reply:
locally in the repo folder run: `pip install tox` and `tox -e py`.
I'll see if I can add this to contributing or similar the next days.

https://github.com/adafruit/workflows-circuitpython-libs/issues/17 should address this in CI ultimately.

Reply:
Since I'll need to move things from `requirements-dev.txt` to `optional_requirements.txt`, I can handle this fix!

Reply:
ah ok. I just pushed a new PR, but it's tiny and I don't mind if you ignore it in favor of yours

Reply:
Ah, no, let's just add it to yours!  Just take all the packages listed in `requirements-dev.txt` and move them to `optional_requirements.txt`.  Then we can delete the unnecessary `requirements-dev.txt`.

Reply:
ok. do I need to maintain the attribution from `requirements-dev.txt`?

Reply:
A good point!  You can add another SPDX line for that author in the merged file so both are retained.

Reply:
cool, pushed an update with those changes as well